RFC/src/RFC-0121_ConsensusEncoding.md

```diff
@@ -0,0 +1,347 @@
+for the signature scalars `u` and `v`.
+
+The encoding is fixed at 96-bytes:
+
+```text
+| 32-byte commitment (R) | 32-byte scalar (u) | 32-byte scalar (v) |
+```
+
+#### Example
+
+Given the following data and types:
+
+```javascript
+{
+  // Type: Fixed array of 5 bytes
+  short_id: [1,2,3,4,5],
+  // Type: variable length string
+  name: "Case",
+  // Type: unsigned integer
+  age: 40,
+  // Type: struct
+  details: {
+      // Type: variable length string
+      kind: "Hacker"
+  },
+  // Type: nullable varint
+  dob: null
+}
+```
+
+Encoded (hex) as follows:
+
+| short id |len|  name  | age |len| kind         | null? | dob |
+|----------|---|--------|-----|---|--------------|-------|-----|
+|0102030405|04 |43617365| 28  |05 | 4861636b6572 | 00    |     |
+
+Note that nested structs are flattened and the order must be preserved to allow decoding.
+The `00` null byte is important so that for e.g. the `kind` bytes cannot be manipulated to 
+produce the same encoding as non-null `dob`.
+
+### Block Header 
+[block header]: #block-header "Block header"
+
+The block hash pre-image is constructed by first constructing the merge mining hash. Each encoding is concatenated in order as follows:
+
+1. `version` - `u16` converted to a fixed 2-byte array (little endian) 
+2. `height` - `u64` converted to a fixed 8-byte array (little endian) 
+3. `prev_hash` - fixed 32-bytes 
+4. `timestamp` - UNIX timestamp converted to a fixed 8-byte array (little endian) 
+5. `input_mr` - fixed 32-bytes
+6. `output_mmr_size` - `u64` converted to a fixed 8-byte array (little endian)
+7. `witness_mr` - fixed 32-bytes
+8. `kernel_mr` - fixed 32-bytes
+9. `kernel_mmr_size` - `u64` converted to a fixed 8-byte array (little endian)
+10. `total_kernel_offset` - 32-byte Scalar, see [RistrettoPrivateKey]
+11. `total_script_offset` - 32-byte Scalar, see [RistrettoPrivateKey]
+
+This pre-image is hashed and block hash is constructed, in order, as follows:
+
+1. `merge_mining_hash` - As above
+2. `pow_algo` - enumeration of types of PoW as a single unsigned byte, where `Monero = 0x00` and `Sha3 = 0x01`
+3. `pow_data` - raw variable bytes (no length varint)
+4. `nonce` - the PoW nonce, `u64` converted to a fixed 8-byte array (little endian)
+
+#### Output Features 
+[Output Features]: #output-metadata-and-features "Output Features"
+
+```rust,ignore
+pub struct OutputFeatures {
+  pub version: OutputFeaturesVersion,
+  pub maturity: u64,
+  pub flags: OutputFlags,
+  pub metadata: Vec<u8>,
+  pub unique_id: Option<Vec<u8>>,
+  pub parent_public_key: Option<PublicKey>,
+  pub asset: Option<AssetOutputFeatures>,
+  pub mint_non_fungible: Option<MintNonFungibleFeatures>,
+  pub sidechain_checkpoint: Option<SideChainCheckpointFeatures>,
+}
+```
+
+Output features consensus encoding is defined as follows (in order):
+
+1. `version` - 1 unsigned byte. This should always be `0x00` but is reserved for future proofing.
```

Comment:
Perhaps a varint rather?

Reply:
I don't have a strong opinion, 256 versions seems like plenty but obviously varint is completely future proof. As long as version < 128 we can change to varint later without it being a breaking change. Maybe the best argument is to have all integers are varint for consistency, but idk.

Reply:
Yeah, it really doesn't matter